receiver/stats.go

```diff
@@ -0,0 +1,126 @@
+/*
+ * skogul, stats receiver
+ *
+ * Copyright (c) 2021 Telenor Norge AS
+ * Author(s):
+ *  - Håkon Solbjørg <hakon.solbjorg@telenor.no>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+
+package receiver
+
+import (
+	"context"
+	"time"
+
+	"github.com/telenornms/skogul"
+	"github.com/telenornms/skogul/stats"
+)
+
+var statsLog = skogul.Logger("receiver", "stats")
+
+// Stats receives metrics from skogul and forwards it to a handler.
+type Stats struct {
+	Handler  *skogul.HandlerRef
+	Interval skogul.Duration
+	ChanSize uint64
+	ch       chan *skogul.Metric
+	ticker   *time.Ticker
+}
+
+// Start starts listening for Skogul stats and
+// emits them on the configured interval.
+func (s *Stats) Start() error {
+	return s.StartC(context.Background())
+}
+
+// StartC allows starting Stats with a context.
+func (s *Stats) StartC(ctx context.Context) error {
+	if s.Interval.Duration == 0 {
+		statsLog.Debugf("Missing interval for stats reporting, defaulting to every %d seconds", stats.DefaultInterval)
+		s.Interval.Duration = stats.DefaultInterval
+	}
+	if s.ChanSize == 0 {
+		s.ChanSize = 100
+	}
+
+	// XXX: we shouldn't allow multiple stats receivers instantiated probably
+	// because they'll just steal the stats from each other.
+	// or we'll have to direct stats to a specific stats instance.
+
+	s.ch = make(chan *skogul.Metric, s.ChanSize)
+
+	s.ticker = time.NewTicker(s.Interval.Duration)
+
+	go s.runner()
+
+	stats.StatsDrainCancel()
+
+	for metric := range stats.Chan {
+		if len(s.ch) >= cap(s.ch) {
+			statsLog.Debug("Dropping stats because the channel is full")
+			continue
+		}
+		select {
+		case s.ch <- metric:
+			continue
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
+// runner is the function listening for stats and emits
+// them when there is time for it.
+func (s *Stats) runner() {
```

Comment:
Ok, I know I've approved this earlier and I think it's fine the way it is, but I think what we want is a different logic.

Right now, metrics are sent by a module, then artificially delayed for no apparent benefit. If the stats were actually modified up until this interval, it would make sense, but right now, I just don't know what the benefit of this algorithm is as opposed to just sending everything as it is received? Sure, it ends up batching more metrics together, which is more performant, but we have a batch sender that'll do that even better.

I feel like we had some idea here for this logic that slipped our mind someway along the road to implementation.

Reply:
My main reasoning for the interval is to keep sending stats even if no new stats have occurred.

Another option is to send stats as soon as they happen *and* periodically. What do you think?

Reply:
Slightly never mind my previous comment. We have two intervals configured: one in the module sending stats and one in the stats-receiver. I agree that we only need one of these. I think it makes the most sense to remove the one here, in "runner", and allow modules to ship stats on their own interval - and skogul will ship them on immediately. If wanted, batchers or other senders can be used to achieve the same benefit.